sdk/core/Azure.Core/src/HttpMessage.cs

```diff
@@ -112,9 +117,41 @@ public bool TryGetProperty(string name, out object? value)
         /// <param name="value">The property value.</param>
         public void SetProperty(string name, object value)
         {
-            _properties ??= new Dictionary<string, object>();
+            _typeProperties ??= new Dictionary<Type, object>();
+            Dictionary<string, object> properties;
+            if (!_typeProperties.TryGetValue(typeof(MessagePropertyKey), out var rawValue))
+            {
+                properties = new Dictionary<string, object>();
+                _typeProperties[typeof(MessagePropertyKey)] = properties;
+            }
+            else
+            {
+                properties = (Dictionary<string, object>)rawValue!;
+            }
+            properties[name] = value;
+        }
 
-            _properties[name] = value;
+        /// <summary>
+        /// Gets a property that modifies the pipeline behavior. Please refer to individual policies documentation on what properties it supports.
+        /// </summary>
+        /// <param name="type">The property type.</param>
+        /// <param name="value">The property value.</param>
+        /// <returns><c>true</c> if property exists, otherwise. <c>false</c>.</returns>
+        internal bool TryGetInternalProperty(Type type, out object? value)
+        {
+            value = null;
+            return _typeProperties?.TryGetValue(type, out value) == true;
```

Comment:
Do you need the `== true`?

Reply:
I think you would need this because it is a `bool?`